SEP-0030: Make compatible with CAP-33 (#744)
stellar/stellar-protocol: ecosystem/sep-0030.md

- The client, authenticated as the external authentication method (e.g. phone, email, etc) or another account, requests the server to sign a transaction.
- The server verifies that the [JWT] is valid according to the authentication provider's specification.
- The server verifies that the external authentication method (e.g. phone, email, etc) or account in the [JWT] is an alternative identity of the account.

> - The server verifies that the source account of the transaction is one of:

Reviewer: ```suggestion
- The server verifies that the source account of the transaction is:
```

Author: I think we need the one of here otherwise it isn't explicit if the list is `and` or `or`. Is this incorrect to have one of here?

Reviewer: You're right, and it is a small change anyway

  - The account the client has authenticated for.
  - In some unique cases an account the server knows definitively for the validity period of the transaction will not be authorized by the signing key.<sup>[1](#note1)</sup>

> - The server verifies that the source account of each operation in the transaction are one of:

Reviewer: ```suggestion
- The server verifies that the source account of each operation in the transaction is:
```

- Find out if a new signing key is available due to signing key rotation.
- Delete accounts.

Notes:

> 1. <a name="note1"></a>See [Endpoints `POST /accounts/<address>/sign/<signing-address>`](#post-accountsaddresssignsigning-address) for more details.

Reviewer: After reading this endpoint description, it is still not completely clear to me how this section provides more details to line 64 and 68. This endpoint appears to sign transactions containing signer adjustments, but its unclear what the explicit connection is.

Is the concern that an account being authorized (made a signer) for another account shouldn't be used as the source account for any operations being described above?

Author: The connection is the paragraph being added about what accounts a server must authorize.

periodically and update the signer on the account to be the most recently
generated signer.

A server must only authorize operations in signed transactions that operate on
the registered account that the client has authenticated for. This is to ensure
that the server does not authorize operations for an account the client is not
authorized. Generally to achieve this the transaction must only be signed if it
contain operations for the authenticated account. The server can verify that
the source account of the transaction and the source account of all operations
in the transaction match the account in the request. In some cases servers may
need to sign transactions that contain other accounts, like a sponsoring
account used with CAP-33. In those cases a server may be implemented to allow
transactions to contain that specific sponsoring account as a source account,
but the server operator must ensure that the signing keys used by the server

> are never signers of the sponsoring account.

Reviewer: I would split the generalized and concrete examples into two different sections. After re-reading this several times I'm starting to get it, but this paragraph leaves the implementor to interpret what other valid operations could have non-request source account.

Could we define exactly what operations could be permitted? Maybe thats too restrictive, but it feels like there are quasi-defined edge cases lingering here.

Author: I agree, I think that is too restrictive. It is also harder to implement and the additional requirements/limitations that would add don't actually help to prevent the problem and the approach isn't applicable to other non-sponsorship use cases that might have the same problem.

I've broken up the large paragraph and reworded some of it to try and be more direct. See 9540af0.

Reviewer: This is much better, thanks.

> are never capable of authorizing operations for the sponsoring account.

Reviewer 2: What exactly is the scenario you are trying to prevent? Is it that a bad user might include a `BeginSponsoringFutureReserves` operation and if the recoverysigner signs the transaction without the sponsoring account's consent, the client can exploit the sponsoring account? Just want to make sure I understand the case you have in mind. If my understanding is correct, when we add the unique signing key feature, this can be prevented, right?

Finally, when you said `authorizing`, partially authorizing also counts, right?

Reviewer: I would add an explicit example of the account sponsorship situation you mention. Otherwise the generic explanation doesn't quite make it clear.

Author: @howardtw If the signing keys of the server were also signing keys of the sponsoring account and the server allowed the signing of transactions for the sponsoring account then any registered account could request signatures for transactions that authorize operations on the sponsoring account, which is not desirable.

If by partially authorizing you mean the signer was a partial signer of the account, then yes, I think this is implied, but I can make this more explicit. Addressed in 50e1334.

@JakeUrban The discussion about a sponsoring account is the example. What did you have in mind that isn't here?

Reviewer: This is resolve from my point of view.
